dvc/repo/__init__.py

```diff
@@ -146,15 +149,15 @@ def _get_database_dir(self, db_name):
 
     def __init__(
         self,
-        root_dir=None,
-        fs=None,
-        rev=None,
-        subrepos=False,
-        uninitialized=False,
-        config=None,
-        url=None,
-        repo_factory=None,
-        scm=None,
+        root_dir: Optional[str] = None,
+        fs: Optional["FileSystem"] = None,
+        rev: Optional[str] = None,
+        subrepos: bool = False,
+        uninitialized: bool = False,
+        config: Optional[DictStrAny] = None,
+        url: Optional[str] = None,
+        repo_factory: Optional[Callable] = None,
+        scm: Optional["Base"] = None,
```

Comment:
`base` does not have all APIs implemented, let's use an union of noscm and git.

```suggestion
        scm: Optional[Union["Git", "NoSCM"]] = None,
```